File: unix/boot/spp/rpp/rpprat/balpar.r
#-h-  balpar			  854  local   12/01/80  15:53:46
# balpar - copy balanced paren string
   include  defs

   subroutine balpar

   character t, token (MAXTOK)
   character gettok, gnbtok

   integer nlpar

   if (gnbtok (token, MAXTOK) != LPAREN) {
      call synerr ("missing left paren.")
      return
      }
   call outstr (token)
   nlpar = 1
   repeat {
      t = gettok (token, MAXTOK)
      if (t == SEMICOL | t == LBRACE | t == RBRACE | t == EOF) {
	 call pbstr (token)
	 break
	 }
      if (t == NEWLINE)      # delete newlines
	 token (1) = EOS
      else if (t == LPAREN)
	 nlpar = nlpar + 1
      else if (t == RPAREN)
	 nlpar = nlpar - 1
      if (t == ALPHA)
	 call squash (token)
      # else nothing special
      call outstr (token)
      } until (nlpar <= 0)

   if (nlpar != 0)
      call synerr ("missing parenthesis in condition.")

   return
   end

File: unix/boot/spp/rpp/rppfor/balpar.f
      subroutine balpar
      integer t, token (100)
      integer gettok, gnbtok
      integer nlpar
      if (.not.(gnbtok (token, 100) .ne. 40))goto 23000
      call synerr (19Hmissing left paren.)
      return
23000 continue
      call outstr (token)
      nlpar = 1
23002 continue
      t = gettok (token, 100)
      if (.not.(t .eq. 59 .or. t .eq. 123 .or. t .eq. 125 .or. t .eq. -1
     *))goto 23005
      call pbstr (token)
      goto 23004
23005 continue
      if (.not.(t .eq. 10))goto 23007
      token (1) = -2
      goto 23008
23007 continue
      if (.not.(t .eq. 40))goto 23009
      nlpar = nlpar + 1
      goto 23010
23009 continue
      if (.not.(t .eq. 41))goto 23011
      nlpar = nlpar - 1
23011 continue
23010 continue
23008 continue
      if (.not.(t .eq. -9))goto 23013
      call squash (token)
23013 continue
      call outstr (token)
23003 if (.not.(nlpar .le. 0))goto 23002
23004 continue
      if (.not.(nlpar .ne. 0))goto 23015
      call synerr (33Hmissing parenthesis in condition.)
23015 continue
      return
      end
